CRM/Event/Form/SelfSvcTransfer.php

```diff
@@ -539,4 +499,59 @@ public function sendCancellation() {
     CRM_Core_Session::setStatus($statusMsg, ts('Thanks'), 'success');
   }
 
+  /**
+   *  Move Participant registration to new contact.
+   *
+   * @param int $toContactID
+   * @param int $fromParticipantID
+   * @param int $eventID
+   *
+   * @throws \CRM_Core_Exception
+   * @throws \CiviCRM_API3_Exception
+   */
+  public function transferParticipantRegistration($toContactID, $fromParticipantID, $eventID) {
+    $query = 'select role_id, source, fee_level, is_test, is_pay_later, fee_amount, discount_id, fee_currency,campaign_id, discount_amount from civicrm_participant where id = ' . $fromParticipantID;
+    $dao = CRM_Core_DAO::executeQuery($query);
+    $value_to = [];
+    while ($dao->fetch()) {
+      $value_to['role_id'] = $dao->role_id;
+      $value_to['source'] = $dao->source;
+      $value_to['fee_level'] = $dao->fee_level;
+      $value_to['is_test'] = $dao->is_test;
+      $value_to['is_pay_later'] = $dao->is_pay_later;
+      $value_to['fee_amount'] = $dao->fee_amount;
+    }
+    $value_to['contact_id'] = $toContactID;
+    $value_to['event_id'] = $eventID;
+    $value_to['status_id'] = CRM_Core_PseudoConstant::getKey(
+      'CRM_Event_BAO_Participant',
+      'status_id',
+      'Registered'
+    );
+    $value_to['register_date'] = date("Y-m-d");
+    //first create the new participant row -don't set registered_by yet or email won't be sent
+    $participant = CRM_Event_BAO_Participant::create($value_to);
+    //send a confirmation email to the new participant
+    $this->participantTransfer($participant);
+    //now update registered_by_id
+    $query = "UPDATE civicrm_participant cp SET cp.registered_by_id = %1 WHERE  cp.id = ({$participant->id})";
+    $params = [1 => [$fromParticipantID, 'Integer']];
+    $dao = CRM_Core_DAO::executeQuery($query, $params);
+    //copy line items to new participant
+    $line_items = CRM_Price_BAO_LineItem::getLineItems($this->_from_participant_id);
```

Comment:
weird to see the test included in this PR is passing without failure as I wrote it to assert the incorrect line items attached to the previous contribution id. On checking, it seems the test passes because the above variable `$this->_from_participant_id` arrives as empty while it is executed. 

Shouldn't it be set to the function parameter `$fromParticipantID`  so that it is also verified in testing process? Or else as this function isn't static anymore, I think we might need to update the test to have a value for $form-> _from_participant_id before calling this function. We might also need to remove the unnecessary params included in this function.

Reply:
@jitendrapurohit the test can be fixed to fail when we add the fix to fix :-)